### PR TITLE
Add arrows to notes

### PR DESCRIPTION
**Pull request by spoorcc**
This pull request solves #117 by adding arrows to notes. Importing of notes from data file should still work. 
I've also cleaned up parts of the Note / NoteModel.

There are some issues that should be solved, but not sure how yet :man_shrugging: 

# Open Issues:
* When clicking note for drag, note jumps
* Arrow position only editable through manage notes dialog
* Arrow position not stored in data file
* Arrow time position, formatting not easy (e.g. filling in just 10 jumps to 00:00:00,000)

Please test before merging :smile:  

demo

### CODE REVIEW

**Comment by spoorcc** [comment]
:see_no_evil:

File: src/customwidgets/notesdockwidget.cpp [hunk 1]
```diff
@@ -69,7 +67,7 @@ void NotesDockWidget::removeNoteRow()
         rows << idx.row();
     }
 
-    qSort(rows.begin(), rows.end(), qGreater<int>());
+    std::sort(rows.begin(), rows.end(), std::greater<int>());
```
Comment (jgeudens) [inline src/customwidgets/notesdockwidget.cpp]: Why use std instead of Qt version?

File: ModbusScope.pro [hunk 2]
```diff
@@ -4,5 +4,5 @@ TEMPLATE = subdirs
 CONFIG += ordered
 
 SUBDIRS += \
-    src \
+    src
```
Comment (jgeudens) [inline ModbusScope.pro]: Why removing " \ ". Isn't it required?

File: src/graphview/basicgraphview.h [hunk 3]
```diff
@@ -90,11 +93,10 @@ protected slots:
     double getClosestPoint(double coordinate);
 
     QVector<QString> _tickLabels;
-    QList<QCPItemText *> _notesItems;
+    QList< std::shared_ptr<NoteItem> > _notesItems;
```
Comment (jgeudens) [inline src/graphview/basicgraphview.h]: Why std::shared_ptr and not Qt version?

File: src/customwidgets/notesdockwidget.cpp (hunk 1 above)
Comment (spoorcc) [reply src/customwidgets/notesdockwidget.cpp]: The Qt ones are deprecated in Qt5:

[source](https://doc.qt.io/qt-5/qtalgorithms.html)

> # Qt and the STL Algorithms
> Historically, Qt used to provide functions which were direct equivalents of many STL algorithmic functions. Starting with Qt 5.0, you are instead encouraged to use directly the implementations available in the STL; most of the Qt ones have been deprecated (although they are still available to keep the old code compiling).

File: src/graphview/basicgraphview.h (hunk 3 above)
Comment (spoorcc) [reply src/graphview/basicgraphview.h]: Forgotten that Qt has its own, reading [this forum post](https://forum.qt.io/topic/19073/any-difference-between-std-shared_ptr-and-qsharedpointer/4) I guess probably best to swap it for qt version

File: ModbusScope.pro (hunk 2 above)
Comment (spoorcc) [reply ModbusScope.pro]: If you don't want to be nagged by the tests it is required :smile:

File: src/graphview/basicgraphview.h (hunk 3 above)
Comment (spoorcc) [reply src/graphview/basicgraphview.h]: Solved in 0162eb8

File: ModbusScope.pro (hunk 2 above)
Comment (spoorcc) [reply ModbusScope.pro]: Fixed in de427d5

**Comment by jgeudens** [comment]
I have rewritten the tests so that they have a separate project file (no more long compilation times for the ModbusScope). I have also added your user manual to the doc folder + start of component overview)
You can rebase on the master.

I will hopefully have this weekend some time to go over your open point and give some suggestions.

**Comment by spoorcc** [comment]
Thanks!

**Comment by jgeudens** [comment]
# _Open Issues_
* _When clicking note for drag, note jumps_
  * I suspects that this has something to do with calculating the offset of the cursor in the note. Does the - operator of QPointF behave like you expect?  https://github.com/spoorcc/ModbusScope/blob/0162eb8ff43a1b8ad204fa53d61e7f53e7d2cb50/src/graphview/basicgraphview.cpp#L488
* _Arrow position only editable through manage notes dialog_
  * I would make this also draggable. Maybe lock the relative position difference between note text and arrow while dragging the arrow?
* _Arrow position not stored in data file_
  * These should be added. The arrow position is the most important value of a note. The note text position isn't that important, although it is nice for the user that the notes are all in the same place after reloading the data file. So save the note text position to the data file and if it is't present that take a guess :).
* _Arrow time position, formatting not easy (e.g. filling in just 10 jumps to 00:00:00,000)_
  * I suggest to make the note text and arrow position read only. They should changed with dragging. Inputting will always be difficult to do.

**Comment by spoorcc** [comment]
pfff, september :see_no_evil:, rebased to restart work

**Comment by jgeudens** [comment]
In the master, I have moved some code from the BasicGraphView to a separate module (GraphViewZoom).
Maybe it is good idea to do the same for the notes code. I think it is feasable.

**Comment by jgeudens** [comment]
I have performed a dirty cherry pick of some generic improvements of the code. Most stuff that I wanted to merge wasn't done in separate commits.

I have done an extensive rework of the dialog code in a branch (will be merged soon). Feel free to rebase and reopen this pull request when you want to finish the feature.

**Pull request closed without merging**